x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -886,7 +884,7 @@ public static String getCreatorRealmName(final Authentication authentication) {
         if (authentication.getAuthenticatedBy().getType().equals(API_KEY_REALM_TYPE)) {
             return (String) authentication.getMetadata().get(API_KEY_CREATOR_REALM_NAME);
         } else {
-            return authentication.getAuthenticatedBy().getName();
+            return authentication.getNominalRealm().getName();
```

Comment:
I think we need to really consider whether this is the change we want.
It's a fairly fundamental change to the ApiKey service, and it affects how the `manage_own_api_key` privilege would work.
It might be right, I'd need to think about it, but if it is, then we might consider it a breaking change.

Reply:
Here is my understanding: 

In generaly, I feel the changed behaviour is correct, though you are right about it is a breaking change.

The above code is only used in `TransportInvalidateApiKeyAction` and `TransportGetApiKeyAction`. The `else` part will be called when the authentication is **not** of ApiKey auth type. Since ApiKey auth cannot perform *runAs*, this means the`else` part will only makes a difference when userA runs as userB.  In this case, the existing code will return userA realm, while the changes will return userB realm. 

I can think of two reasons to support the changed behaviour:
* When apikey is created, it records the runas user (if there is one) and runas/lookedUpBy realm. It feels correct that the same runas/lookedUpBy info should be checked when retrieve/invalidate the apikey. 
* When you run as someone else, all permissions and stuff should be checked/recorded as someone else. 

I don't quite understand how `manage_own_api_key` work, the document says 
```
All security-related operations on Elasticsearch API keys that are 
owned by the current authenticated user. The operations include 
creating new API keys, retrieving information about API keys, 
and invalidating API keys.
```
Does this mean you will not be able to create an API key for yourself unless you have this priviledge? If this is the case, I don't see how this is related to the change. Could you please elaborate on it?

Reply:
Tim saved the day, I completely missed this 😞 .

@yang I think what you're describing is mostly correct.

> Does this mean you will not be able to create an API key for yourself unless you have this priviledge?

Yes

> If this is the case, I don't see how this is related to the change. Could you please elaborate on it?

Let's say you run-as and create an API key. The user that you run as only has the `manage_own_api_key` privilege. The creation would work. But then invalidation (and get), also using the run-as, won't work because the API Key Service would search the APi Key as incorrectly belonging to a user in the realm that authenticated the caller, whereas the API Key is created under the run-as user. (Note, I haven't actually tested this scenario).

In the interest of this PR, I would revert to the previous behavior in the `TransportInvalidateApiKeyAction` and `TransportGetApiKeyAction` , and, assuming you agree this is a bug, you should open a bug issue with your finding.

Reply:
I tried to simulate the scenario you described with two users `foo` and `bar`, where `foo` is a superuser and `bar` only has `manage_own_api_key` for cluster and read for all indices.  ES version 7.5.2

It behaves pretty much as you described when requests are made with `owner=true`. Given foo creates an ApiKey while runas bar, the request
`curl -u foo:password -H 'es-security-runas-user: bar' 'localhost:9200/_security/api_key?id=keyId&owner=true'` returns empty results.

While I do believe the correct behaviour of above request should be returning the api key, I also think it is better to have a separate PR for this as it is not related to the current issue.

PS: I also found if `bar` has only `manage_own_api_key` priviledge, it is not possible to get or delete the api key. I got error says `action [cluster:admin/xpack/security/api_key/invalidate] is unauthorized for user [bar]`. Once I add `manage_api_key` to `bar`, everything works.

Reply:
I [reverted the logic](https://github.com/elastic/elasticsearch/pull/51305/files#diff-cc4f9d4c564aae71214ac709ed789b4dR887) and created #51975 to track it. Thanks

Reply:
Thank you Yang! But currently I don't think we want to set the authenticating realm in the case of the set security user processor, see my other comment: https://github.com/elastic/elasticsearch/pull/51305/files?file-filters%5B%5D=.java#r375923902

Reply:
Agreed. Please see my [other comment](https://github.com/elastic/elasticsearch/pull/51305#discussion_r376124488).